### PR TITLE
Distributor optim

### PR DESCRIPTION
**Pull request by tardyp**
WIP fix for #3661

### CODE REVIEW

File: master/buildbot/newsfragments/getOldersBuildrequest.bugfix [hunk 1]
```diff
@@ -0,0 +1 @@
+Improve ``getOldestRequestTimeOptim`` function of buildrequest distributor to to sorting and paging in the database layer.
```
Comment (seankelly) [inline master/buildbot/newsfragments/getOldersBuildrequest.bugfix]: Doubled "to". Is the second one supposed to be "do" instead?

File: master/buildbot/process/buildrequestdistributor.py [hunk 2]
```diff
@@ -472,21 +473,28 @@ def _activityLoop(self):
...
 
-            # bail out if we shouldn't keep looping
-            if not self.running or not self._pending_builders:
+                # bail out if we shouldn't keep looping
+                if not self._pending_builders:
+                    self.pending_builders_lock.release()
+                    self.activity_lock.release()
+                    break
+                # take that builder list, and run it until the end
+                pending_builders = copy.copy(self._pending_builders)
```
Comment (seankelly) [inline master/buildbot/process/buildrequestdistributor.py]: Why the copy?

Comment (tardyp) [reply master/buildbot/process/buildrequestdistributor.py]: If you look at how locks are setup, it is not guaranteed that pending_builders will not be messed up while we are running the loop. 

Current implementation is not doing it, as it is playing with set and sorts. The locks looks useless as far as I can read. But as this stuff is pretty central, I prefer to be overcautious.

**Comment by tardyp** [comment]
included in #3684

**Pull request closed without merging**